### ISSUE TITLE
[css-nesting] How to resolve nested CSS with pseudo elements in the parent

### ISSUE THREAD

**Issue by romainmenke** [w3c/csswg-drafts#7433]
first reported here : https://github.com/csstools/postcss-plugins/issues/510

relevant specification parts: 

https://www.w3.org/TR/css-nesting-1/#nest-selector

>The [nesting selector](https://www.w3.org/TR/css-nesting-1/#nesting-selector) can be desugared by replacing it with the parent style rule’s selector, wrapped in an [:is()](https://www.w3.org/TR/selectors-4/#matches-pseudo) selector. For example,

https://www.w3.org/TR/selectors-4/#matches-pseudo

>Pseudo-elements cannot be represented by the matches-any pseudo-class; they are not valid within [:is()](https://www.w3.org/TR/selectors-4/#matches-pseudo).

For this CSS :

```pcss
.anything::before {
  @nest .something_else > & {
    color: black
  }
}
```

`&` should represent the matched element of the enclosing rule. Which in this case is `before`.
As far as I know we can not express `.something_else > before` at the moment.

Is this CSS :
1. invalid
2. matches `.something_else > .anything::before`
3. matches `.something_else.anything :> before`  (borrowing `:>` as a child pseudo element combinator)

Also relevant : 

- https://github.com/w3c/csswg-drafts/issues/2284#issuecomment-364580632
- https://github.com/w3c/csswg-drafts/issues/7346#issue-1266265523

**Comment by romainmenke** [comment]
cc @argyleink @tabatkins

**Comment by argyleink** [comment]
I think you did the right thing reverting the change that tried to correct it for developers. I'm thinking wait/lean on devtools to help developers learn that :`is()` and `:where()` can't accept pseudo elements? add it to a style linter to watch out for? or the postcss plugin could notice nested `::` and warn in the console? 

agree that this is a tough lesson in nesting, since the 1st lesson to be learned with `is` and `where` feels out of no where. heh, `no:where()`.. 

i believe this is how the nested selectors can be rewritten. 

```css
.anything {
  &::before {}

  @nest .something_else > &::before {
    color: black
  }
}
```

this makes me think we could add a warning to the spec that any nesting inside a `::before` (or of same type) will produce an invalid selector per https://www.w3.org/TR/selectors-4/#matches-pseudo and have no matches.

**Comment by romainmenke** [comment]
Adding a warning to the spec would be great! Anything that is invalid in `:is` is also invalid as a nesting ancestor (ancestor of the rule, not of the matched element)

At the moment this is implied but not clear beyond any possible confusion.

It's true that with nesting specifically stylesheet authors need to re-learn the feature and this is a painful process :/

It was inspired by sass nesting but it is completely different from that.
The first versions of the PostCSS plugin did something different altogether.

So stylesheet authors have an intuitive sense of how the feature should work that doesn't match how it will eventually work in browsers.

We already took some steps with the last major version of the PostCSS plugin but clearly more work is needed :)

------

A (style)lint rule that can spot invalid nesting would be awesome and doesn't exist yet.
This also gives faster feedback while typing.

**Comment by cdoublev** [comment]
I am not sure I understand why `::before { &:hover {} }` would be invalid. I do not claim it should. Even `:is()` not accepting pseudo-elements is not something already defined, I think. But I'm reading the note about the `:is()` desugaring trick as intended only for CSS processing tools authors.

**Comment by cdoublev** [comment]
I was checking if `&div` is valid and found this example in [3. Nesting Selector: the & selector](https://drafts.csswg.org/css-nesting-1/#nest-selector):

> For example, `&div` is a valid nesting selector, meaning "whatever the parent rules matches, but only if it’s also a `<div>` element".

If I understand this correctly, `div { &div {} }` would be (silly but) valid and you cannot desugar it with `:is(div)div`. I am not sure to understand why `&` compounded with `<type-selector>` would be valid, though. It may be an actual bug in the spec.

**Comment by johannesodland** [comment]
> Adding a warning to the spec would be great! Anything that is invalid in `:is` is also invalid as a nesting ancestor (ancestor of the rule, not of the matched element)

As an author, I hope we don't drop nesting inside pseudo-elements just because it's not supported in :is() today. 
There are many use cases for nesting in pseudo-elements, and there will be more as more nested pseudo-elements are added.

**Nested conditional rules**
```css
div::before {
  color: blue;

  @media (min-width: 480px) {
    color: red;
  }
}
```

**Nested user action pseudo classes **
```css
div::before {
  color: blue;

  &:hover {
    color: red;
  }
}
```

**Nested pseudo-elements**
```css
article::fist-letter {
  initial-letter: 3;

  &::prefix {
    font-size: 0.5em;
    vertical-align: top;
  }
}
```

It's not just that we are used to the existing feature. Making nesting inside pseudo-elements invalid will block many use cases, and will make it harder to teach and understand nesting. The specificity side-effects of using :is() will be surprising enough as it is.

I understand making it invalid would make the feature easier to implement through :is() at the moment. I hope we can prioritise authors over ease of implementation, and perhaps look at extending support for nested pseudo-elements in :is(). Improved syntax for selecting pseudo-element children and descendants as described in [7346](https://github.com/w3c/csswg-drafts/issues/7346) would be great too.

**Comment by tabatkins** [comment]
The :is() desugaring, taken literally, just advice for tools, yeah. We don't *actually* produce an :is() internally. But the restrictions should exist in both places or neither.

**Comment by sesse** [comment]
FWIW, while the implementation in Blink doesn't literally produce an :is() internally (it couldn't, due to CSSOM demands), it produces something that is awfully close. It goes through the exact same path while matching, for instance. (See https://chromium-review.googlesource.com/c/chromium/src/+/3934883/16/third_party/blink/renderer/core/css/selector_checker.cc#1315 if you're interested)

**Comment by romainmenke** [comment]
TL;DR;

`@supports selector(:is(::before)) and selector(& p)` is a valid way to check for support in the future if this remains true :

> But the restrictions should exist in both places or neither.

----------

I was looking at this from a feature detection perspective and I think a couple of things intersect here.

- `:is()` is not forgiving in `@supports` which was a recent change I think
- you can not test for a structure in `@supports` (e.g. `@supports selector(::before { & {} })`)


Chrome 106:
- `@supports selector(:is(::before))` : `true`
- `@supports selector(& p)` : `false`
- `@supports selector(:is(::before)) and selector(& p)` : `false`

Chrome 109:
- `@supports selector(:is(::before))` : `false`
- `@supports selector(& p)` : `true`
- `@supports selector(:is(::before)) and selector(& p)` : `false`

```html
<!DOCTYPE html>
<html lang="en">
<head>
	<title>Document</title>
	<style>
		html {
			& p {
				color: green;
			}
		}

		@supports selector(& p) {
			p {
				/* Version 109.0.5372.0 (Official Build) canary (arm64) */
				/* Experimental platform features on */
				text-decoration: underline solid green;
			}
		}

		@supports selector(:is(::before)) {
			p {
				/* Version 106.0.5249.119 (Official Build) (arm64) */
				text-decoration: underline wavy red;
			}
		}

		@supports selector(:is(::before)) and selector(& p) {
			p {
				text-decoration: underline dotted purple;
			}
		}
	</style>
</head>
<body>
	<p>Hello!</p>
</body>
</html>
```

**Comment by LeaVerou** [comment]
Conceptually, there is no reason to disallow this:

```pcss
.anything::before {
  .something_else > & {
    color: black
  }
}
```

So it would be unfortunate if we need to for what is basically an implementation detail. Perhaps we can just change how selectors get rewritten?

**Comment by romainmenke** [comment]
Yes, but your comment hints at a larger change.

```css
.parent .a, .other-parent .b::before {
	.c > & {
		color: black
	}
}
```

currently matches like :

```css
.c > :is(.parent .a), .c > :is(.other-parent .b::before) {
	color: black
}
```

Which is also why I agree with @tabatkins to keep these linked.
Resolving this issue for both nesting and `:is()` at the same time is the least confusing.

**Comment by johannesodland** [comment]
If I’m not mistaken, precompilers used to fully expand nested selectors. I’m not sure if it was a huge problem?

If nested rules were fully expanded, we could rely on linters to warn authors of the expanded complexity. :is() can then be added selectively where needed. 

I think I might prefer this over nesting not working inside pseudo-elements and the unexpected specificity. This would also need linting to warn about invalid use.

**Comment by romainmenke** [comment]
preprocessor tools come in roughly two categories : `scss`-like nesting and those that attempt to follow the specification.

`scss`-like nesting doesn't have this issue at all because they manipulate selector bits as strings. `.foo { &--bar }` -> `.foo--bar {}`

Those that attempt to follow the specification had/have bugs and deviations.
A recent attempt to align [postcss-nesting](https://github.com/csstools/postcss-plugins/tree/main/plugins/postcss-nesting#readme) with the specification is why I opened this issue in the first place.

No tool that I know of ever fully expanded nested selectors as would be required to follow the specification.

> I think I might prefer this over nesting not working inside pseudo-elements and the unexpected specificity.

I think everyone agrees that this should work and that it is more a matter of finding the right way to fix this.

**Comment by sesse** [comment]
> I think everyone agrees that this should work

I don't agree this should be solved for css-nesting-1. I'm fine with some followup spec (be it css-nesting or css-selectors) making :is() more flexible (thus trickling down to nesting), as long as the edge cases are clearly worked out.

**Comment by romainmenke** [comment]
> I don't agree this should be solved for css-nesting-1.

I agree, my intention was to clarify that not fixing it now (in css-nesting-1) doesn't mean it won't every be updated.

**Comment by LeaVerou** [comment]
> 3 just raises the question of why we can't do that stuff in `:is()`.

Indeed, why?

**Comment by Loirooriol** [comment]
> Conceptually, there is no reason to disallow this:
> ```css
> .anything::before { .something_else > & { color: black  }}
> ```

@LeaVerou I disagree: the reason is that `::before` pseudo-elements cannot be children (in the DOM).

`::` is like a combinator, so these are different:
```css
.something_else > .anything::before { color: black  }
.anything::before { .something_else > & { color: black  }}
```
Just like these are different:
```css
.something_else > .anything > before { color: black  }
.anything > before { .something_else > & { color: black  }}
```

**Comment by LeaVerou** [comment]
@Loirooriol `::` is not like a combinator. You cannot rearrange things like you can for combinators, `::before` is a single token. In the second example, the `.something else` and `.anything` can be the same element in the nested selected, but not in the non-nested one. However, no such thing is true for the former.

**Comment by Loirooriol** [comment]
Strictly speaking is not a combinator but it kinda behaves like a combinator and probably should have been a combinator and #7346 resolved to work on actual combinators for accessing pseudo-elements.

I don't get your rearranging argument, `.foo > .bar > .baz` can't be rearranged either as `.foo > .baz > .bar` or something, it's normal not being able to rearrange complex selectors. That's actually an argument in favor of `::` behaving kinda like a combinator.

**Comment by romainmenke** [comment]
> That's actually an argument in favor of :: behaving kinda like a combinator.

yup : `.foo::before:hover` vs. `.foo:hover::before`

**Comment by LeaVerou** [comment]
> Strictly speaking is not a combinator but it kinda behaves like a combinator and probably should have been a combinator and #7346 resolved to work on actual combinators for accessing pseudo-elements.
> 
> I don't get your rearranging argument, `.foo > .bar > .baz` can't be rearranged either as `.foo > .baz > .bar` or something, it's normal not being able to rearrange complex selectors. That's actually an argument in favor of `::` behaving kinda like a combinator.

I didn't mean that kind of rearranging, but that with combinators you can do things like `.foo > :is(.bar) > .baz`, adjust whitespace etc because they are separate tokens. I'm still missing the explanation on how your pseudo-element example differs between the nested and non-nested version.

**Comment by tabatkins** [comment]
> If I’m not mistaken, precompilers used to fully expand nested selectors. I’m not sure if it was a huge problem?

They did not, precisely because this *is* a huge problem. Sass uses some heuristics to guess what subset of the possible selectors to actually emit. I presume other langs do so as well (or just have a combinatorial explosion in their output sometimes).

> I don't agree this should be solved for css-nesting-1. I'm fine with some followup spec (be it css-nesting or css-selectors) making :is() more flexible (thus trickling down to nesting), as long as the edge cases are clearly worked out.

Yup, that's my intent as well.

**Comment by Loirooriol** [comment]
@LeaVerou The difference is that `.something_else > .anything::before` means: select the `before` pseudo-elements originated by an element with class `anything` which is a child of an element with class `something_else`.

And `.anything::before { .something_else > &` means: select the things which are children of an element with class `something_else` if they also happen to be `before` pseudo-elements originated by an element with class `anything`. And it's not possible for a `before` pseudo-element to be a child of an element in the DOM (since the pseudo-element doesn't exist in the DOM), so this won't match anything.

I think `.anything::before { &:hover {` seems doable (basically a matter of letting `&` select whatever matches the parent selector, regardless of whether it's an element or a pseudo-element. But I don't see how `.anything::before { .something_else > &` can work in a sane way without fully expanding the selectors.

**Comment by johannesodland** [comment]
> They did not, precisely because this _is_ a huge problem.

I’m sorry, my mental model was off. I didn’t see the full complexity. 

Adding an explanation of why `.anything::before { .something_else > &` is invalid to the spec would be helpful. 

Linting and devtools will help, and hopefully nesting in pseudo-elements / a more flexible :is() can be solved in a later spec 🤞🏼

**Comment by Loirooriol** [comment]
> Fully expand the selectors, possibly resulting in a combinatorial explosion of selector length. (.a .b .c { .d .e & {...}} expands to, hm, 5-choose-2 selectors (10 total, in this case).

Not that it matters much, but the explosion is worse than that, because the simple selectors can refer to the same element: `.a .b .c .d .e`, `.a .b .c.d .e`, `.a .b .c .d .e`, `.a .b.d .c .e`... I think that's 25 in total. So yeah, fully expanding is bad.

**Comment by tabatkins** [comment]
Right, I allowed for that in my calculation - that's why it's 5-choose-2. Note that the expanded selector has to end with `.c` on its own; you're just merging the `.a .b` and `.d .e` preludes together, and there are five locations in the first selector (each compound, and between/before/after each) where you can drop in the two compounds of the second selector, so 5-choose-2 combos in total.

**Comment by tabatkins** [comment]
I now have an explicit restriction against & representing pseudo-elements in the spec, along with an issue noting that we'd like to fix it, but need to do so simultaneously for :is() since they rely on the same mechanisms.

**Comment by sesse** [comment]
Thanks!

**Comment by cdoublev** [comment]
> we'd like to fix it, but need to do so simultaneously for :is() since they rely on the **same mechanisms**.

Do you mean the *match a selector* algorithms? (I am not familiar with them yet)

**Comment by matthew-dean** [comment]
@tabatkins

I know this is an old conversation, but just FWIW:

> They did not, precisely because this is a huge problem. Sass uses some heuristics to guess what subset of the possible selectors to actually emit. I presume other langs do so as well (or just have a combinatorial explosion in their output sometimes).

Less essentially clamps the number of combinations you can have. Otherwise it's a terrible problem, and at least in Less threads / documentation, we warn authors that this is one of the dangers of over-nesting lists.

Interestingly, I was going to open a suggestion in Less to, at some point, output `:is()` wrappers instead of combining selectors. Although this may not be necessary since it could just pass through nesting.

**Comment by matthew-dean** [comment]
> Interestingly, I was going to open a suggestion in Less to, at some point, output :is() wrappers instead of combining selectors.

Hang on, pseudo-elements are not valid in `:is()`? 🤔  Why?

**Comment by tabatkins** [comment]
Because an element is never a pseudo-element (and you can't generically select a pseudo-element and then test its name in an `:is()`).

That is, `div:is(::before)` is asking "is the `div` a `::before`?" and that's never true (and similarly, `::before:is(::before)` is either always or never true, depending on if the names match or not). What people *expect* it to mean is "change the subject of the selector to the div's ::before pseudo", but pseudo-classes can't change the subject; you need a combinator for that. 

Ultimately the problem is that early CSS screwed up the pseudo-element syntax, making it look like a pseudo-class (literally spelled it `:before`; the double-colon came later) rather than like a combinator+name. (There's some other issues to potentially fix that, tho I can't dig them up at the moment.)

**Comment by yisibl** [comment]
This limitation is sure to confuse more and more authors, even developers of browser engines, see https://issues.chromium.org/issues/40278599

**Comment by LeaVerou** [comment]
@yisibl can you please post this in 9492 which is the actual open issue for this?

**Comment by yisibl** [comment]
> @yisibl can you please post this in 9492 which is the actual open issue for this?

No problem. Done

**Comment by rctneil** [comment]
Just wanted to post that I have recently hit an issue which (i think) this is the culprit of. I'd love to see this scenario relaxed as it confused me for a good long time.

Styles I was trying to use:

```
::part(form-control-label) {
    .filter-bar & {
        color: yellow;
    }
}
```

Discussion here: https://discord.com/channels/739437527907303535/1020332605599662181/1346201308448686191 and here: https://discord.com/channels/436251713830125568/1345141673478328532
